src/main/scala/model/world/Reproduction.scala

```diff
@@ -26,19 +27,29 @@ object Reproduction {
    * @param dad another bunny
    * @return the 4 children of the couple, one for each cell of the Punnett's square
    */
-  def generateChildren(mom: Bunny, dad: Bunny): Population = {
+  def generateChildren(mom: Bunny, dad: Bunny, mutations: Option[List[Mutation]] = None): Population = {
     var childrenGenotypes = List.fill(CHILDREN_EACH_COUPLE)(PartialGenotype(Map()))
+
     Genes.values.foreach(gk => {
-      val grandmaMomAllele = mom.genotype(gk).momAllele
-      val grandpaMomAllele = mom.genotype(gk).dadAllele
-      val grandmaDadAllele = dad.genotype(gk).momAllele
-      val grandpaDadAllele = dad.genotype(gk).dadAllele
-      val anotherGene = Random.shuffle(
-        List( Gene(gk, grandmaMomAllele, grandmaDadAllele),
-              Gene(gk, grandpaMomAllele, grandmaDadAllele),
-              Gene(gk, grandmaMomAllele, grandpaDadAllele),
-              Gene(gk, grandpaMomAllele, grandpaDadAllele)))
-      childrenGenotypes = (for (i <- 0 until CHILDREN_EACH_COUPLE) yield childrenGenotypes(i) + anotherGene(i)).toList
+      val grandmaMomAllele = StandardAllele(mom.genotype(gk).momAllele.kind)
+      val grandpaMomAllele = StandardAllele(mom.genotype(gk).dadAllele.kind)
+      val grandmaDadAllele = StandardAllele(dad.genotype(gk).momAllele.kind)
+      val grandpaDadAllele = StandardAllele(dad.genotype(gk).dadAllele.kind)
+      val genesOfReproduction : List[Gene]= List( Gene(gk, grandmaMomAllele, grandmaDadAllele),
+                                                  Gene(gk, grandpaMomAllele, grandmaDadAllele),
+                                                  Gene(gk, grandmaMomAllele, grandpaDadAllele),
+                                                  Gene(gk, grandpaMomAllele, grandpaDadAllele) )
+      val anotherGene = Random.shuffle(genesOfReproduction)
+
+      mutations match {
+        case None => childrenGenotypes = (for (i <- 0 until CHILDREN_EACH_COUPLE) yield childrenGenotypes(i) + anotherGene(i)).toList
+        case Some(ms) => ms filter(_.geneKind == gk) foreach {
+          _ => childrenGenotypes = (for (i <- 0 until CHILDREN_EACH_COUPLE - 1) yield childrenGenotypes(i) + anotherGene(i)).toList
+          Gene(gk, JustMutatedAllele(gk.mutated), JustMutatedAllele(gk.mutated)) :: childrenGenotypes
```

Comment:
Secondo me in questo case ci va `_ => childrenGenotypes = (for (i <- 0 until CHILDREN_EACH_COUPLE - 1) yield childrenGenotypes(i) + anotherGene(i)).toList :: Gene(gk, JustMutatedAllele(gk.mutated), JustMutatedAllele(gk.mutated))` altrimenti l'ultimo Gene, quello mutato, non l'aggiunge veramente a childrenGenotypes

Reply:
Non è un case, è il parametro preso in input dal foreach, solo che è stato scritto sotto.

Reply:
Allora non capisco a cosa serva il foreach, dovrebbero esserci 0 o 1 mutazione per quel gene e se ce ne sono 0 dobbiamo comportarci come nel caso sopra...

caso `None` o mutazione non di questo genekind => `childrenGenotypes = (for (i <- 0 until CHILDREN_EACH_COUPLE) yield childrenGenotypes(i) + anotherGene(i)).toList`
caso mutazione di questo genekind => `childrenGenotypes = Gene(gk, JustMutatedAllele(gk.mutated), JustMutatedAllele(gk.mutated)) :: (for (i <- 0 until CHILDREN_EACH_COUPLE - 1) yield childrenGenotypes(i) + anotherGene(i)).toList`

Reply:
si la filter ritornerà sempre o una lista vuota o al massimo una lista con un elemento che è la mutazione corrispondente a quel genekind. Sono passata attraverso il foreach per una questione di tipi dato che con un apply non mi corrispondevano più i tipi di ritorno che nel nostro cosa sono una lista di PartialGenotype mentre la apply richiede un intero come tipo di ritorno. 
Ho deciso quindi di aggiungere inizialmente tutti e quattro i bunny e nel caso si trovi la mutazione si toglie l'ultimo e si sostituisce con il nuovo mutato